safekeeper/src/wal_backup_partial.rs

```diff
@@ -270,8 +275,32 @@ impl PartialBackup {
     }
 }
 
+/// Check if everything is uploaded and partial backup task doesn't need to run.
+pub(crate) fn needs_uploading(
+    state: &StateSnapshot,
+    uploaded: &Option<PartialRemoteSegment>,
+) -> bool {
+    match uploaded {
+        Some(uploaded) => {
+            uploaded.status != UploadStatus::Uploaded
+                || uploaded.flush_lsn != state.flush_lsn
+                || uploaded.commit_lsn != state.commit_lsn
+                || uploaded.term != state.term
```

Comment:
Was thinking about using `term` vs `last_log_term` briefly, but didn't know the issue is possible, nice find!

Ok, so the fix is to change partial backup to always use `last_log_term` instead of `term`. Then it will automatically reupload all partial segments if `term` and `last_log_term` mismatch. Does it sound good?


> Also upload_segment should probably have a comment like ''s3 file might not really contain the data of its name as we do upload without lock and recheck term later; only Uploaded status in cfile guarantees data is correct".

I don't fully understand the reasoning, but will add more comments to `UploadStatus` to highlight that only `Uploaded` state contains correct data.

Reply:
Changed to using `last_log_term` everywhere.